accepted/0000-link-packages-to-source-and-build.md

```diff
@@ -0,0 +1,514 @@
+- Running the signed timestamp authority for Rekor as well as a Certificate Transparency monitor for Fulcio. This will help spread the trust outside the four walls of Sigstore.
+- Maintaining the Sigstore trust root and having a root key holder as part of the group of Sigstore root key holders.
+
+We also remain open to an alternative solution if Sigstore is not able to meet npm's requirements for uptime and support.
+
+## Detailed Explanation
+### Sigstore as a solution to package signing
+Package signing is a mitigation for package hijacking attacks. If maintainers sign their packages, end-users can verify that packages were produced by their corresponding owners. However, as of today, the main existing solution involves maintainer-managed PGP (or similar) keys, which has known [usability](https://www.usenix.org/legacy/events/sec99/full_papers/whitten/whitten.ps) [and](https://latacora.micro.blog/2019/07/16/the-pgp-problem.html) [security](https://blog.cryptographyengineering.com/2014/08/13/whats-matter-with-pgp/) [issues](https://www.wired.co.uk/article/efail-pgp-vulnerability-outlook-thunderbird-smime): registries that have implemented public key signing have seen either limited adoption or achieved adoption only by mandating it (e.g., Maven Central, which reports developer friction).
+
+[Sigstore](https://www.sigstore.dev/) is a promising alternative: it allows signing artifacts with short-lived “disposable/ephemeral keys” notarized against log-ins with [OpenID Connect](https://openid.net/connect/) (OIDC) identities (e.g., Google or GitHub accounts).
+
+Several package ecosystems have announced interest in adopting Sigstore as a way for maintainers to sign packages (e.g., PyPI, RubyGems, Maven Central). [OpenSSF](https://blog.chainguard.dev/sigstore-statement-for-openssf-call/) is also working to standardize on Sigstore for signing software.
+
+However, adopting something like Sigstore raises several major questions (beyond the risks detailed above):
+- What security benefits does Sigstore provide?
+- How should maintainers authenticate to sign packages? The solution should provide both high assurance in the authenticity of packages and respect maintainer privacy.
+
+#### What security benefits does Sigstore provide?
+[Sigstore](https://www.sigstore.dev/) is a collection of tools and services aimed at making it easy to use short-lived, ephemeral certificates to sign software.
+
+Sigstore has three main components: a CLI tool (Cosign), a Certificate Authority ([fulcio.sigstore.dev](https://docs.sigstore.dev/fulcio/overview/)), and a time stamping and immutable ledger service ([rekor.sigstore.dev](https://docs.sigstore.dev/rekor/overview/)).
+
+The Cosign functionality will be embedded directly in npm CLI, so we're left with the public [fulcio.sigstore.dev](https://docs.sigstore.dev/fulcio/overview/) (CA) and [rekor.sigstore.dev](https://docs.sigstore.dev/rekor/overview/) (ledger) services.
```

Comment:
As someone unfamiliar with this ledger: Are there any concerns around exponential growth of this ledger and the insertion of new entries slowing down publishing? Other.... ledgers that I'm aware of begin to hit speed issues that would be concerning at the volume of _all published packages_.

Additionally, when you say `The Cosign functionality will be embedded directly in npm CLI` does this mean that the CLI will reach out to the Fulcio/Rekor APIs, or that additional binaries would be included?

Reply:
The ledger is sharded - Every year a new tree is created, so this should alleviate concerns around performance. Additionally, while not currently implemented, we'd like to have log mirrors to distribute the load, but this will require the community operating mirrors, which comes with performance expectations and storage costs.

> The Cosign functionality will be embedded directly in npm CLI

I assume this is referring to performing the same behavior as Cosign, using the sigstore-js library to call out to Rekor/Fulcio.

Reply:
> Every year a new tree is created, so this should alleviate concerns around performance.

I'd be really curious to know what the throughput is. npm has... a lot of publishes, and I get concerned about even a yearly reset if this will also include other ecosystems that are also growing + new ecosystems that haven't yet been invented.

What happens when the equivalent of all packages in 2022 are published in one week? 😅

Reply:
It will likely take a very long time before all daily package publishes are signing their builds. Assuming the uptick will be gradual over time.

The [database](https://github.com/google/trillian) that backs the transparency logs of both Rekor and Fulcio has very high throughput. It's the same database that powers the [certificate transparency log](https://transparency.dev/ ) for all SSL certs being issued and performs over 2,000 writes per second.

Reply:
> It will likely take a very long time before all daily package publishes are signing their builds. Assuming the uptick will be gradual over time.

IIRC npm's growth over time hasn't necessarily been gradual 😅 

> The [database](https://github.com/google/trillian) that backs the transparency logs of both Rekor and Fulcio has very high throughput. It's the same database that powers the [certificate transparency log](https://transparency.dev/) for all SSL certs being issued and performs over 2,000 writes per second.

IMO it would be really nice to do some of the math before this RFC is ratified to see how much runway we have until we hit 2,000 writes per second across all the ecosystems that have begun considering this. It'd be good to know that this is/isn't a problem and how far out we need to start thinking about being concerned :)

Reply:
I agree that some envelope arithmetic would be ideal. My current estimate is that it'll be no big deal: 2k RPS is 2.88M per day. From searching a bit I see counts of ~2M packages in npm. I doubt that every package is signed every day, so there's some headroom.

The 2k probably isn't a hard upper limit either. [Let's Encrypt reported early last year on their current DB servers](https://letsencrypt.org/2021/01/21/next-gen-database-servers.html) -- 2x EPYC giving 64 cores, with 2TB RAM. It's possible to buy [4 socket EPYC systems with up to 4TB of RAM](https://www.broadberry.com/superservers-supermicro-servers/sys-2014tp-htr).